scheduler/src/cook/mesos/mesos_mock.clj

```diff
@@ -0,0 +1,430 @@
+    (when (some nil? offers)
+      (throw (ex-info "Unknown offers"
+                      {:unknown-offers
+                       (remove #(contains? (set (keys offer-id->offer)) %)
+                               offer-ids)})))
+    (when-not (every? #(= slave-id (:value (:slave-id %))) offers)
+      (throw (ex-info "Some offers don't match slave id of tasks"
+                      {:expected-slave-id slave-id
+                       :mismatched-offers (filter #(not= slave-id (:value (:slave-id %)))
+                                                  offers)})))
+    (let [host (get slave-id->host slave-id)
+          available-resources (combine-resources (mapcat :resources offers) (:available-resources host))
+          requested-resources (combine-resources (mapcat :resources tasks))
+          resources' (subtract-resources available-resources requested-resources)
+          tasks (map #(assoc % :launched-time (t/now)) tasks)
+          task->runtime-ms (-> state :config :task->runtime-ms)]
+      (log/info "Resources requested by tasks: " {:requested-resources requested-resources})
+      ;; May need to put this in a thread..
+      (doseq [task tasks]
+        (.statusUpdate scheduler
+                       driver
+                       (mesos-type/->pb :TaskStatus
+                                        {:task-id {:value (:value (:task-id task))}
+                                         :state :task-running})))
+      (log/debug "Launching tasks " {:tasks tasks})
+      (-> state
+          (update :task-id->task #(into % (map-from-vals (comp :value :task-id) tasks)))
+          (update :complete-chan->task-id
+                  #(into % (map (fn [{:keys [task-id] :as task}]
+                                  [(async/timeout (task->runtime-ms task)) (:value task-id)]))
+                         tasks))
+          (assoc-in [:slave-id->host slave-id :available-resources] resources')
+          (update :offer-id->offer #(apply dissoc % offer-ids))))))
+
+(defmethod handle-action! :reconcile
+  [action statuses {:keys [offer-id->offer task-id->task slave-id->host] :as state} driver scheduler]
+  ;; TODO: implement
+  )
+
+(defn default-task->runtime-ms
+  "Takes a task spec and casts the command to a number
+   as runtime"
+  [task]
+  (-> task :command :value read-string))
+
+(defn default-task->complete-status
+  "Returns completed successfully"
+  [_]
+  ;; TODO: support more completion states
+  :task-finished)
+
+(defn mesos-driver-mock
+  "A mock mesos implementation.
+
+   It stores the 'hosts' in the cluster, how much resources are available on
+   each host, what jobs are running on each 'host' and how much longer each
+   job is expected to run for.
+
+   Parameters:
+   `hosts` is a seq of maps that contain a hostname, agent id, resources and attributes
+   `speed-multiplier` is a number >1 which is how much to speed up 'time'
```

Comment:
The speed multiplier isn't really used yet but the intent with all of this is to write a faster than real time simulator where we can play back what happened in a production cluster in order to test improvements to the scheduling algorithm. The fact that it helps us improve our testing is an added bonus.

In the simulator, being about to pick the speed up factor is important to compare different runs of the simulation.